File: js/sdk/src/v3/clients/prompts.ts
import { r2rClient } from "../../r2rClient";
import {
  WrappedBooleanResponse,
  WrappedGenericMessageResponse,
  WrappedPromptResponse,
  WrappedPromptsResponse,
} from "../../types";

export class PromptsClient {
  constructor(private client: r2rClient) {}

  /**
   * Create a new prompt with the given configuration.
   *
   * This endpoint allows superusers to create a new prompt with a
   * specified name, template, and input types.
   * @param name The name of the prompt
   * @param template The template string for the prompt
   * @param inputTypes A dictionary mapping input names to their types
   * @returns
   */
  async create(options: {
    name: string;
    template: string;
    inputTypes: Record<string, string>;
  }): Promise<WrappedGenericMessageResponse> {
    return this.client.makeRequest("POST", "prompts", {
      data: options,
    });
  }

  /**
   * List all available prompts.
   *
   * This endpoint retrieves a list of all prompts in the system.
   * Only superusers can access this endpoint.
   * @returns
   */
  async list(): Promise<WrappedPromptsResponse> {
    return this.client.makeRequest("GET", "prompts");
  }

  /**
   * Get a specific prompt by name, optionally with inputs and override.
   *
   * This endpoint retrieves a specific prompt and allows for optional
   * inputs and template override.
   * Only superusers can access this endpoint.
   * @param options
   * @returns
   */
  async retrieve(options: {
    name: string;
    inputs?: string[];
    promptOverride?: string;
  }): Promise<WrappedPromptResponse> {
    const data: Record<string, any> = {
      ...(options.inputs && { inputs: options.inputs }),
      ...(options.promptOverride && {
        promptOverride: options.promptOverride,
      }),
    };

    return this.client.makeRequest("POST", `prompts/${options.name}`, {
      params: data,
    });
  }

  /**
   * Update an existing prompt's template and/or input types.
   *
   * This endpoint allows superusers to update the template and input types of an existing prompt.
   * @param options
   * @returns
   */
  async update(options: {
    name: string;
    template?: string;
    inputTypes?: Record<string, string>;
  }): Promise<WrappedGenericMessageResponse> {
    const params: Record<string, any> = {
      name: options.name,
    };
    if (options.template) {
      params.template = options.template;
    }
    if (options.inputTypes) {
      params.inputTypes = options.inputTypes;
    }

    return this.client.makeRequest("PUT", `prompts/${options.name}`, {
      data: params,
    });
  }

  /**
   * Delete a prompt by name.
   *
   * This endpoint allows superusers to delete an existing prompt.
   * @param name The name of the prompt to delete
   * @returns
   */
  async delete(options: { name: string }): Promise<WrappedBooleanResponse> {
    return this.client.makeRequest("DELETE", `prompts/${options.name}`);
  }
}

File: py/sdk/asnyc_methods/prompts.py
import json
from typing import Any, Optional

from shared.api.models import (
    WrappedBooleanResponse,
    WrappedGenericMessageResponse,
    WrappedPromptResponse,
    WrappedPromptsResponse,
)


class PromptsSDK:
    def __init__(self, client):
        self.client = client

    async def create(
        self, name: str, template: str, input_types: dict
    ) -> WrappedGenericMessageResponse:
        """Create a new prompt.

        Args:
            name (str): The name of the prompt
            template (str): The template string for the prompt
            input_types (dict): A dictionary mapping input names to their types
        Returns:
            dict: Created prompt information
        """
        data: dict[str, Any] = {
            "name": name,
            "template": template,
            "input_types": input_types,
        }
        response_dict = await self.client._make_request(
            "POST",
            "prompts",
            json=data,
            version="v3",
        )

        return WrappedGenericMessageResponse(**response_dict)

    async def list(self) -> WrappedPromptsResponse:
        """List all available prompts.

        Returns:
            dict: List of all available prompts
        """
        response_dict = await self.client._make_request(
            "GET",
            "prompts",
            version="v3",
        )

        return WrappedPromptsResponse(**response_dict)

    async def retrieve(
        self,
        name: str,
        inputs: Optional[dict] = None,
        prompt_override: Optional[str] = None,
    ) -> WrappedPromptResponse:
        """Get a specific prompt by name, optionally with inputs and override.

        Args:
            name (str): The name of the prompt to retrieve
            inputs (Optional[dict]): JSON-encoded inputs for the prompt
            prompt_override (Optional[str]): An override for the prompt template
        Returns:
            dict: The requested prompt with applied inputs and/or override
        """
        params = {}
        if inputs:
            params["inputs"] = json.dumps(inputs)
        if prompt_override:
            params["prompt_override"] = prompt_override
        response_dict = await self.client._make_request(
            "POST",
            f"prompts/{name}",
            params=params,
            version="v3",
        )

        return WrappedPromptResponse(**response_dict)

    async def update(
        self,
        name: str,
        template: Optional[str] = None,
        input_types: Optional[dict] = None,
    ) -> WrappedGenericMessageResponse:
        """Update an existing prompt's template and/or input types.

        Args:
            name (str): The name of the prompt to update
            template (Optional[str]): The updated template string for the prompt
            input_types (Optional[dict]): The updated dictionary mapping input names to their types
        Returns:
            dict: The updated prompt details
        """
        data: dict = {}
        if template:
            data["template"] = template
        if input_types:
            data["input_types"] = json.dumps(input_types)
        response_dict = await self.client._make_request(
            "PUT",
            f"prompts/{name}",
            json=data,
            version="v3",
        )

        return WrappedGenericMessageResponse(**response_dict)

    async def delete(self, name: str) -> WrappedBooleanResponse:
        """Delete a prompt by name.

        Args:
            name (str): The name of the prompt to delete
        Returns:
            bool: True if deletion was successful
        """
        response_dict = await self.client._make_request(
            "DELETE",
            f"prompts/{name}",
            version="v3",
        )

        return WrappedBooleanResponse(**response_dict)
